README.md
metricsd
============

I am a node.js client library for
[metricsd](https://github.com/mojodna/metricsd), a metrics aggregator for
[Graphite](http://graphite.wikidot.com) that supports counters, histograms and
meters. Metricsd 0.4.2+ is recommended.

[![Build
Status](https://secure.travis-ci.org/mojodna/node-metricsd.png?branch=master)](http://travis-ci.org/mojodna/node-metricsd)

Install with `npm install metricsd` or download if you prefer.


metricsd?
=========

[Yeah](https://github.com/mojodna/metricsd). It's like
[statsd](https://github.com/etsy/statsd), but with more types and less heavy
lifting, and thus an expanded protocol. `statsd` clients are compatible with
`metricsd`, but not vice versa. Sorry. If you wanted to, you could probably
implement a compatibility mode.


Usage
=====

The library currently exports a factory function that creates an instance of
a Metrics object with the given `options`:

```javascript
// make a metrics instance (default options are shown)
var metricsd = require('metricsd'),
    metrics = metricsd({
        host: "localhost",
        port: 8125,
        enabled: true,
        log: false,
        logger: console.log,
        prefix: null,
        timeout: 1000
    });
```

The available options are:

* `host` and `port` - server settings.
* `enabled` - set to `false` if you don't want to send metrics while testing.
* `prefix` - if you run more than one environment, machine or module, supply
  a `prefix` to identify all metrics from this instance.
* `log` - whether to log metrics (using `logger`) instead of sending them using
  a socket.
* `logger` - `console.log`-compatible logger.
* `timeout` - node-metricsd cleans up the internal socket if it's idle using
  this timeout (milliseconds), and periodically cleans up `Buffer`s using
  `10 * timeout` (a work-around for 0.6.7 and earlier).
* `socket` - you may provide your own dgram socket. If so, `timeout` is
  ignored.

The `metrics` instance exposes the options above as properties: all are
read-only with the exception of `enabled` which may be toggled at any time.


Counters
========

Create a new `Counter` with `metrics.count(name)` to track relative values.

```javascript
  var counter = metrics.count('numThings');
  counter.inc(2);  // +2
  counter.dec();   // -1
```

Alternatively, write directly to a named counter with convenience functions on
`metrics` itself:

* `metrics.count(name)` - create a counter
* `metrics.inc(name, value)` - update a counter by +value
* `metrics.dec(name, value)` - update a counter by -value
* `metrics.updateCounter(name, value)` - update a counter by +value
* `metrics.deleteCounter(name)` - tell metricsd to stop tracking a counter


Gauges
======

Create a new `Gauge` with `metrics.gauge(name)` to track absolute values that
are obtained on a regular basis.

```javascript
var gauge = metrics.gauge('numThings');
gauge.update(10); // numThings == 10
gauge.update(20); // numThings == 20
gauge.update(5);  // numThings == 5
```

Alternatively, write directly to a named gauge with convenience functions on
`metrics` itself:

* `metrics.gauge(name)` - create a gauge
* `metrics.updateGauge(name, value)` - set the named gauge's value
* `metrics.deleteGauge(name)` - tell metricsd to stop tracking a gauge


Histograms
==========

Create a new `Histogram` with `metrics.histogram(name)` to track intermittent
values and their statistical breakdowns (max, min, mean, median, 75th
percentile, etc.).  (metricsd does the hard work.)

```javascript
var histogram = metrics.histogram('numThings');
histogram.update(10); // min, max, mean == 10
histogram.update(20); // min == 10, max == 20, mean == 10
histogram.update(5);  // min == 5, max == 10, mean == 17.5
```

Alternately, write directly to a named histogram with convenience functions on
`metrics` itself:

* `metrics.histogram(name)` - create a histogram
* `metrics.updateHistogram(name, value)` - set the named histogram's value
* `metrics.deleteHistogram(name)` - tell metricsd to stop tracking a histogram


Timers
======

Create a new `Timer` with `metrics.time(name)`:

```javascript
var timer = metrics.time('thingTime');

setTimeout(function() {
  timer.stop();             // thingTime == 10ms (approx)
  assert.ok(timer.stopped); // time should be stopped
}, 10);
```

Or time a series of related events with `metrics.time(name)` and
`timer.lap(name)`:

```javascript
var timer = metrics.time('otherTime');

setTimeout(function() {
  timer.lap('otherTime.lap1'); // flushes otherTime.lap1 == 10ms (approx)
  assert.ok(timer.running);    // time should not be stopped
}, 10);

setTimeout(function() {
  timer.lap('otherTime.lap2'); // flushes otherTime.lap2 == 10ms (approx)
  assert.ok(timer.running);    // time should not be stopped

  var duration = timer.stop(); // flushes otherTime == 20ms (approx)
  assert.ok(timer.stopped);    // time should be stopped
}, 20);
```

You can also pause and resume timers (e.g. if you don't want to include time
spent "waiting"):

```javascript
var timer = metricstime('sporadic');

timer.pause();

setTimeout(function() {
  timer.resume();

  var duration = timer.stop(); // flushes sporadic == 0ms (approx)
}, 10);
```

Or time an event until a callback with `metrics.time(name,callback)`. Method
arguments are passed on untouched:

```javascript
var callback = metrics.timeCallback('thingTime', function(a, b, c) {
  console.log('thing happened %d,%d,%d', a, b, c); // thing happened 1,2,3
});

// ... 123ms later
callback(1,2,3); // thingTime == 123ms
```

Timers start automatically, but can be restarted by calling `timer.start()`.
You can wrap callbacks yourself with `timer.wrap(callback)`.

To delete a timer, use `metrics.deleteHistogram(name)`. To update a statistical
measurement directly, use `metrics.updateHistogram(name, value)`.


Meters / Marks
==============

To mark the occurrence of a named event, use `metrics.mark(name)`. This is
useful for measuring event rates (e.g. requests per second).

To tell metricsd to stop tracking a meter, use `metrics.deleteMeter(name)`.


Raw Metrics
===========

To send a raw metric with no formatting or prefixes applied, use
`metrics.write(metric)`. This is used internally by all other metrics. See
[metricsd](https://github.com/mojodna/metricsd) for more info on formatting.


Console Output
==============

Specify `log: true` to emit metrics to `STDOUT` rather than the default UDP
socket.

If you provide a `logger` to the `metricsd()` factory (`console.log` or
compatible), it will be used to emit metric strings in the form `metric:
<metric>`. This can be convenient if your logs streams are being consumed by
something that can more efficiently funnel data into `metricsd` or similar.

```javascript
var metricsd = require('metricsd'),
    metrics = metricsd({
      log: true,
      logger: console.log
    });
```

License
=======

Copyright (c) 2012-2013 Seth Fitzsimmons and Tom Carden

Published under the MIT License.


HISTORY.md
vNext
=====

* Pass the elapsed time as the last argument when using `metrics.timeCallback`

v0.2.2: Jul 18 2013
===================

* Fixed formatting breakage introduced in 0.2.1

v0.2.1: Jul 18 2013
===================

* `metrics.format(fmt, args...)` suitable for use with `metrics.write()`

v0.2.0: Jul 11 2013
===================

* High-resolution timing (using `process.hrtime()`)
* `Timer.pause()` and `Timer.resume()`
* Log stream output (as `metric=<metric>`)
* Cleaned up socket handling; `close()` is no longer necessary when using Node
  v0.10.x+
* Removed batched metrics

v0.1.4: Apr 18 2013
===================

* Added missing `metrics.histogram()`

v0.1.3: Mar 6 2012
==================

* Stop generating invalid metric values when batched

v0.1.2: Mar 6 2012
==================

* Limit batch sizes to 8kb to avoid UDP send failures

v0.1.1: Mar 6 2012
==================

* Express-compatible middleware (`metrics.middleware`)
* Batch flushing of metrics

v0.1.0: Feb 26 2012
===================

* Initial public release


test/middleware.js
"use strict";

var EventEmitter = require("events").EventEmitter,
    metricsd = require("../lib/metricsd"),
    expect = require("chai").expect;

describe("metrics.middleware", function() {
    var metrics;
    var _send;
    var req;
    var res;
    var next;

    beforeEach(function() {
        metrics = metricsd();

        // allow metrics._send() to be mocked
        _send = metrics._send;

        req = {};
        res = new EventEmitter();
        next = function() {};
    });

    afterEach(function() {
        metrics._send = _send;

        _send = undefined;

        req = undefined;
        res = undefined;
        next = undefined;
    });

    describe(".Measure", function() {
        var event;
        var metric;

        beforeEach(function() {
            event = "finished";
            metric = "requestTime";
        });

        afterEach(function() {
            event = undefined;
            metric = undefined;
        });

        it("should return Express-compatible middleware", function(done) {
            var fn = metrics.middleware.Measure(event, metric);

            fn(req, res, done);
        });

        it("should register an event handler for the specified event name", function(done) {
            var fn = metrics.middleware.Measure(event, metric);

            fn(req, res, function() {
                expect(res.listeners(event)).to.have.length(1);

                done();
            });
        });

        it("should update the named histogram when the specified event fires", function(done) {
            var fn = metrics.middleware.Measure(event, metric);

            fn(req, res, next);

            metrics._send = function(str) {
                expect(str).to.match(new RegExp(metric + ":\\d+\\|h"));

                done();
            };

            res.emit(event);
        });

        it("should treat the 'name' parameter as a name lookup if it's a function", function(done) {
            var fn = metrics.middleware.Measure(event, function(req, res, eventName, elapsedTime) {
                expect(eventName).to.equal(event);

                return metric;
            });

            fn(req, res, next);

            metrics._send = function(str) {
                expect(str).to.match(new RegExp(metric + ":\\d+\\|h"));

                done();
            };

            res.emit(event);
        });

        it("should handle lookups returning multiple names", function(done) {
            var metricNames = [metric, "test.requestTime"];

            var fn = metrics.middleware.Measure(event, function(req, res, eventName, elapsedTime) {
                return metricNames;
            });

            fn(req, res, next);

            var sent = 0;
            metrics._send = function(str) {
                expect(str).to.match(new RegExp(metricNames[sent++] + ":\\d+\\|h"));

                if (sent === 2) {
                    done();
                }
            };

            res.emit(event);
        });
    });

    describe(".TimeFilter", function() {
        var metric;

        beforeEach(function() {
            metric = "providedMiddleware";
        });

        afterEach(function() {
            metric = undefined;
        });

        it("should be Express-compatible middleware", function(done) {
            var fn = metrics.middleware.TimeFilter(metric, function(req, res, next) {
                next();
            });

            fn(req, res, function() {
                done();
            });
        });

        it("should call the provided filter", function(done) {
            var fn = metrics.middleware.TimeFilter(metric, function(req, res, next) {
                done();
            });

            fn(req, res, next);
        });

        it("should measure the time that the provided filter takes", function(done) {
            var fn = metrics.middleware.TimeFilter(metric, function(req, res, next) {
                next();
            });

            metrics._send = function(str) {
                expect(str).to.match(new RegExp(metric + ":\\d+\\|h"));

                done();
            };

            fn(req, res, next);
        });
    });

    describe(".Timer", function() {
        it("should be Express-compatible middleware", function(done) {
            metrics.middleware.Timer(req, res, done);
        });

        it("should add a 'timer' property to the provided request", function(done) {
            metrics.middleware.Timer(req, res, function() {
                expect(req.timer).to.be.an.instanceof(metrics.Timer);

                done();
            });
        });

        it("should leave an existing 'timer' property alone", function(done) {
            var timer = {};
            req.timer = timer;

            metrics.middleware.Timer(req, res, function() {
                expect(req.timer).to.equal(timer);

                done();
            });
        });
    });
});


lib/middleware.js
"use strict";

/**
 * Factory function intended to be called by the `middleware` getter in
 * metricsd.
 *
 * @param {Object} metrics Configured metricsd instance.
 */
module.exports = function(metrics) {
    var middleware = {};

    /**
     * Measure the time until an event (on the response) fires. It will use the
     * request timer (added via middleware.Timer) if available, otherwise it
     * will create one. If you'd like accurate timings, ensure that
     * middleware.Timer (or this) is as early in the filter chain as possible.
     *
     * @param {String} eventName Response event.
     * @param {String|Array|Function} metricName Metric(s) to write; if
     * a lookup function, it will be called with (req, res, eventName,
     * elapsedTime) and is expected to return a String or an Array.
     *
     * @return {Function} An Express-compatible middleware function.
     */
    middleware.Measure = function(eventName, metricName) {
        return function(req, res, next) {
            middleware.Timer(req, res, function() {
                res.once(eventName, function() {
                    var elapsedTime = req.timer.elapsedTime;

                    if (metricName instanceof Function) {
                        metricName = metricName(req, res, eventName, elapsedTime);
                    }

                    var metricNames = [metricName];
                    if (metricName instanceof Array) {
                        metricNames = metricName;
                    }

                    metricNames.forEach(function(name) {
                        metrics.updateHistogram(name, Math.round(elapsedTime));
                    });
                });

                next();
            });
        };
    };

    /**
     * Measure the time the provided middleware function takes. For example:
     *
     * Instead of:
     *   app.use(express.cookieParser());
     * to measure its duration:
     *   app.use(metrics.timeMiddleware('express.cookieParser', express.cookieParser()));
     *
     * @param {String} metricName Metric name to write.
     * @param {Function} fn Middleware function to measure.
     *
     * @return {Function} An Express-compatible middleware function.
     */
    middleware.TimeFilter = function(metricName, fn) {
        return function(req, res, next) {
            var timer = metrics.time();

            fn(req, res, function() {
                timer.stop(metricName);

                next.apply(this, arguments);
            });
        };
    };

    /**
     * Establish a timer that can be used throughout the duration of the
     * request / response cycle. This _is_ Express-compatible middleware and
     * should be added to your filter chain as early as possible to ensure
     * correct timings. 
     *
     * @param {Object} req Request.
     * @param {Object} res Response.
     * @param {Function} next Next filter in the chain.
     */
    middleware.Timer = function(req, res, next) {
        req.timer = req.timer || metrics.time();

        next();
    };

    return middleware;
};


lib/metricsd.js
"use strict";

var util = require("util");

/**
 * Export a factory function to create a Metrics instance.
 *
 * Default options are:
 *
 * {
 *     host: "localhost",
 *     port: 8125,
 *     enabled: true,
 *     log: false,
 *     logger: console.log,
 *     prefix: null,
 *     timeout: 1000
 * }
 *
 * You may also provide your own dgram-ish socket as options.socket if required.
 *
 * If you don't provide a socket, the internal socket will be closed to free
 * up resources every after the metrics instance has been idle for at least
 * options.timeout milliseconds. A delay of 10*timeout is also used to close
 * the internal socket in order to clear Buffers it has been holding on to.
 *
 * If options.enabled is false, all code will run but API.write will never emit
 * metrics (either via socket or log), and therefore the socket will never be
 * opened. Good for testing locally if you don't care about metrics there.
 */
module.exports = function(options) {
    options = options || {};

    var host = options.host || "localhost";
    var port = options.port || 8125;
    var providedSocket = options.socket || null;
    var prefix = options.prefix || null;
    var enabled = 'enabled' in options ? options.enabled : true;
    var socketTimeout = options.timeout || 1000;
    var log = 'log' in options ? options.log : false;
    var logger = options.logger || console.log;

    var ephemeralSocket = null;
    var lastUse = null;
    var gcTimer = null;
    var closeTimeout = null;
    var middleware = null;

    /**
     * Close the ephemeral socket.
     */
    var closeSocket = function() {
        if (ephemeralSocket) {
            ephemeralSocket.close();
            ephemeralSocket = null;

            // this was probably called due to closeTimeout, but clear it just
            // in case
            clearTimeout(closeTimeout);
            closeTimeout = null;
        }
    };

    /**
     * Clear the socket cleanup timer.
     */
    var removeTimer = function() {
        if (gcTimer) {
            clearInterval(gcTimer);
            gcTimer = null;
        }
    };

    /**
     * Garbage collect the socket if possible.
     */
    var gcSocket = function() {
        if (Date.now() - lastUse > socketTimeout) {
            closeSocket();
            removeTimer();
        }
    };

    var getSocket = function() {
        if (providedSocket) {
            return providedSocket;
        } else {
            // create an ephemeral socket
            if (!ephemeralSocket) {
                ephemeralSocket = require("dgram").createSocket("udp4");

                // don't wait for this to be closed
                ephemeralSocket.unref && ephemeralSocket.unref();

                // prevent errors (that we don't care about) from propagating
                ephemeralSocket.on("error", function(err) {});

                // try to clean up the socket periodically to free up resources
                // if this instance is idle
                gcTimer = setInterval(gcSocket, 250);

                // don't wait for the timer to exit
                gcTimer.unref && gcTimer.unref();

                // forcibly close the socket periodically (ignoring last use)
                // to allow sent Buffers to be GC'd
                closeTimeout = setTimeout(closeSocket, 10 * socketTimeout);

                // don't wait for the timer to exit
                closeTimeout.unref && closeTimeout.unref();
            }

            // the ephemeral socket was last used NOW.
            lastUse = Date.now();

            return ephemeralSocket;
        }
    };

    // object that will be returned by require("metrics")()
    var API = {};

    /**
     * Release resources created by the metrics library.
     */
    API.close = function() {
        closeSocket();
        removeTimer();
    };

    Object.defineProperty(API, "enabled", {
        get: function() {
            return enabled;
        },
        set: function(value) {
            enabled = value;
        },
        enumerable: true
    });

    Object.defineProperty(API, "host", {
        get: function() {
            return host;
        },
        enumerable: true
    });

    Object.defineProperty(API, "log", {
        get: function() {
            return log;
        },
        set: function(value) {
            log = value;
        },
        enumerable: true
    });

    Object.defineProperty(API, "logger", {
        get: function() {
            return logger;
        },
        set: function(value) {
            logger = value;
        },
        enumerable: true
    });

    Object.defineProperty(API, "port", {
        get: function() {
            return port;
        },
        enumerable: true
    });

    Object.defineProperty(API, "prefix", {
        get: function() {
            return prefix;
        },
        enumerable: true
    });

    Object.defineProperty(API, "timeout", {
        get: function() {
            return socketTimeout;
        },
        enumerable: true
    });

    Object.defineProperty(API, "socket", {
        get: function() {
            return providedSocket;
        },
        enumerable: true
    });

    Object.defineProperty(API, "middleware", {
        get: function() {
            if (!middleware) {
                middleware = require("./middleware")(this);
            }

            return middleware;
        },
        enumerable: false
    });

    /**
     * The name will be appended to prefix if one was specified in
     * options and whitespace in name will be replaced with underscores.
     */
    var formatMetricName = function(name) {
        if (name) {
            if (prefix) {
                name = prefix + "." + name;
            } else {
                name = name;
            }

            return name.replace(/\s/g, "_");
        }
    };

    API._send = function(str) {
        if (log) {
            logger("metric=%s", str);
            return;
        }

        // append trailing newlines if necessary
        if (str.lastIndexOf("\n") !== str.length - 1) {
            str = str + "\n";
        }

        var buf = new Buffer(str);

        getSocket().send(buf, 0, buf.length, port, host, function(err, bytes) {
            if (err) {
                // console.log("Error while sending data:", err.message);
            }
        });
    };

    API.write = function(metric) {
        if (!enabled) {
            return;
        }

        API._send(metric);
    };

    var format = function(fmt, args) {
        if (!args || args.length === 0) {
            if (fmt instanceof Object) {
                if (Object.keys(fmt).length === 0) {
                    return;
                }

                fmt = Array.prototype.slice.call(fmt);
                return format.apply(null, [fmt[0], fmt.slice(1)]);
            }

            return formatMetricName(fmt);
        }

        // convert an arguments object into an array
        args = Array.prototype.slice.call(args);

        var prefixed = false;

        if (args[0] instanceof Object) {
            if (args[0].length > 0) {
                args[0] = format.apply(null, args[0]);
                prefixed = true;
            } else {
                args[0] = "";
            }
        }

        // determine the number of placeholders present in order to format the
        // string properly (without including the value)
        var placeholders = ((args[0] || "").match(/%\w/g) || "").length;

        // format the name
        var name = util.format.apply(null, args.slice(0, placeholders + 1));

        if (!prefixed) {
            name = formatMetricName(name);
            prefixed = true;
        }

        // format the metric
        return util.format.apply(null, [fmt, name].concat(args.slice(placeholders + 1)));
    };

    API.format = function(fmt, args) {
        args = Array.prototype.slice.call(arguments, 1);

        return format.apply(null, [fmt, args]);
    };

    API.deleteCounter = function(name) {
        API.write(format("%s:delete|c", arguments));
    };

    API.deleteGauge = function(name) {
        API.write(format("%s:delete|g", arguments));
    };

    API.deleteHistogram = function(name) {
        API.write(format("%s:delete|h", arguments));
    };

    API.deleteMeter = function(name) {
        API.write(format("%s:delete", arguments));
    };

    API.updateCounter = function(name, value) {
        // TODO sample rate
        API.write(format("%s:%s|c", arguments));
    };

    API.updateGauge = function(name, value) {
        API.write(format("%s:%s|g", arguments));
    };

    API.updateHistogram = function(name, value) {
        // TODO sample rate
        API.write(format("%s:%s|h", arguments));
    };

    /**
     * Create a named counter.
     *
     * Can be incremented or decremented, which can be useful if you don't have
     * an absolute value to hand. If you do, it's probably better to use
     * a gauge rather than worry about how to initialize a counter correctly ;)
     */
    var Counter = function(name) {
        name = format(arguments);

        if (!name) {
            throw new Error("a name is required");
        }

        if (!(this instanceof Counter)) {
            return new Counter(name);
        }

        this.name = name;
    };

    /**
     * Increment the counter.
     */
    Counter.prototype.inc = function(value) {
        value = value || 1;

        API.updateCounter(this.name, value);
    };

    /**
     * Decrement the counter.
     */
    Counter.prototype.dec = function(value) {
        value = (value && 0 - value) || -1;

        API.updateCounter(this.name, value);
    };

    /**
     * Tell metricsd to stop tracking this counter.
     */
    Counter.prototype.delete = function() {
        API.deleteCounter(this.name);
    };

    /**
     * Create a named gauge.
     *
     * For measuring a continuous value, such as current queue or database
     * size.
     */
    var Gauge = function(name) {
        name = format(arguments);

        if (!name) {
            throw new Error("a name is required");
        }

        if (!(this instanceof Gauge)) {
            return new Gauge(name);
        }

        this.name = name;
    };

    /**
     * Tell metricsd to stop tracking this gauge.
     */
    Gauge.prototype.delete = function() {
        API.deleteGauge(this.name);
    };

    /**
     * Update the gauge's value.
     */
    Gauge.prototype.update = function(value) {
        API.updateGauge(this.name, value);
    };

    /**
     * Create a named histogram.
     */
    var Histogram = function(name) {
        name = format(arguments);

        if (!name) {
            throw new Error("a name is required");
        }

        if (!(this instanceof Histogram)) {
            return new Histogram(name);
        }

        this.name = name;
    };

    /**
     * Tell metricsd to stop tracking this histogram.
     */
    Histogram.prototype.delete = function() {
        API.deleteHistogram(this.name);
    };

    /**
     * Update the histogram's value.
     */
    Histogram.prototype.update = function(value) {
        API.updateHistogram(this.name, value);
    };

    /**
     * Create a named Meter.
     */
    var Meter = function(name) {
        name = format(arguments);

        if (!name) {
            throw new Error("a name is required");
        }

        if (!(this instanceof Meter)) {
            return new Meter(name);
        }

        this.name = name;
    };

    /**
     * Tell metricsd to stop tracking this meter.
     */
    Meter.prototype.delete = function() {
        API.deleteMeter(this.name);
    };

    /**
     * Mark an occurrence of the events that this meter is tracking.
     */
    Meter.prototype.mark = function() {
        API.mark(this.name);
    };

    /**
     * Convert high-resolution time to fractional milliseconds.
     */
    var hrtimeToMS = function(hrtime) {
        return (hrtime[0] * 1e9 + hrtime[1]) / 1e6;
    };

    /**
     * Create a named timer and start it immediately. If you don't wish to start
     * timing immediately, call start() when you're ready.
     */
    var Timer = function(name) {
        name = format(arguments);

        if (!(this instanceof Timer)) {
            return new Timer(name);
        }

        this.name = name;
        this.startTime = null;
        this.lapStartTime = null;
        this.laps = [];
        this.stopped = false;
        this.paused = false;
        this.pauseTime = null;
        this.pausedDuration = 0;

        Object.defineProperty(this, "elapsedTime", {
            get: function() {
                return hrtimeToMS(process.hrtime(this.startTime));
            },
            enumerable: true
        });

        Object.defineProperty(this, "running", {
            get: function() {
                return !this.stopped;
            },
            enumerable: true
        });

        this.start();
    };

    /**
     * Measure a lap time.
     */
    Timer.prototype.lap = function(name) {
        name = format(arguments);

        var lapTime = process.hrtime(this.lapStartTime);
        this.resetLapTimer();

        this.laps.push({
            name: name || "",
            time: hrtimeToMS(lapTime)
        });

        if (name) {
            API.updateHistogram(name, Math.round(hrtimeToMS(lapTime)));
        }

        return hrtimeToMS(lapTime);
    };

    /**
     * Reset the lap timer.
     */
    Timer.prototype.resetLapTimer = function() {
        this.lapStartTime = process.hrtime();
    };

    /**
     * Start (or restart) the timer.
     */
    Timer.prototype.start = function() {
        this.startTime = process.hrtime();
        this.lapStartTime = this.startTime;
        this.stopped = false;
        this.paused = false;
        this.pauseTime = null;
        this.pausedDuration = 0;
    };

    /**
     * Stop the timer.
     */
    Timer.prototype.stop = function(name) {
        name = format(arguments) || this.name;

        // repeated calls to stop should do nothing
        if (!this.stopped) {
            this.stopped = true;
            var elapsed = Array.isArray(this.startTime) ? hrtimeToMS(process.hrtime(this.startTime)) : 0;

            this.resume(); // update pausedDuration if paused

            elapsed -= this.pausedDuration;

            if (name) {
                API.updateHistogram(name, Math.round(elapsed));
            }

            return elapsed;
        }
    };

    /**
     * Pause the timer.
     */
    Timer.prototype.pause = function() {
        if (!this.paused) {
            this.paused = true;
            this.pauseTime = process.hrtime();

            return hrtimeToMS(this.pauseTime) - hrtimeToMS(this.startTime) - this.pausedDuration;
        }
    };

    /**
     * Resume the timer.
     *
     * @return Number time spent paused.
     */
    Timer.prototype.resume = function() {
        if (this.paused) {
            this.paused = false;

            var elapsed = hrtimeToMS(process.hrtime(this.pauseTime));

            this.pausedDuration += elapsed;

            return elapsed;
        }
    };

    /**
     * Wrap a callback such that the timer stops when the callback is triggered.
     */
    Timer.prototype.wrap = function(callback) {
        var self = this;

        return function() {
            var time = self.stop();

            var args = Array.prototype.slice.call(arguments);
            args.push(time);

            callback.apply(this, args);
        };
    };

    API.Counter = Counter;
    API.Timer = Timer;
    API.Gauge = Gauge;
    API.Histogram = Histogram;
    API.Meter = Meter;

    /**
     * Create a named counter.
     */
    API.count = function(name) {
        try {
            return new Counter(format(name));
        } catch (e) {}
    };

    /**
     * Create a named gauge.
     */
    API.gauge = function(name) {
        try {
            return Gauge.apply(null, arguments);
        } catch (e) {}
    };

    /**
     * Create a named histogram.
     */
    API.histogram = function(name) {
        try {
            return Histogram.apply(null, arguments);
        } catch (e) {}
    };

    /**
     * Create a named meter.
     */
    API.meter = function(name) {
        try {
            return Meter.apply(null, arguments);
        } catch (e) {}
    };

    /**
     * Increment a counter.
     */
    API.inc = function(name, value) {
        if (arguments.length > 0) {
            value = value || 1;

            API.updateCounter(name, value);
        }
    };

    /**
     * Decrement a counter.
     */
    API.dec = function(name, value) {
        if (arguments.length > 0) {
            value = (value && 0 - value) || -1;

            API.updateCounter(name, value);
        }
    };

    /**
     * For measuring the rate of occurrences of the named event.
     */
    API.mark = function(name) {
        name = format(arguments);

        if (name) {
            API.write(name);
        }
    };

    /**
     * Create a timer with optional name.
     *
     * For measuring the duration of the named event.
     * Name can also be specified in the optional stop() method.
     * Or you can read the elapsedTime property and manually update a histogram
     * yourself.
     */
    API.time = function(name) {
        return Timer.apply(null, arguments);
    };

    /**
     * Time from now until the callback fires.
     *
     * Returns a wrapped callback that you should pass to the method that
     * you're timing. Name and callback are both required.
     */
    API.timeCallback = function(name, callback) {
        if (!name || name instanceof Function) {
            throw new Error("a name is required");
        }

        var args = Array.prototype.slice.apply(arguments);

        // don't pass the function into the timer
        var timer = Timer.apply(null, args.slice(0, -1));
        callback = timer.wrap(callback);
        timer.start();

        return callback;
    };

    return API;
};
